src/librustc_trans/trans/attributes.rs

```diff
@@ -265,7 +265,7 @@ pub fn from_fn_type<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>, fn_type: ty::Ty<'tcx
                 // on memory dependencies rather than pointer equality
                 let interior_unsafe = mt.ty.type_contents(ccx.tcx()).interior_unsafe();
 
-                if mt.mutbl == hir::MutMutable || !interior_unsafe {
+                if mt.mutbl != hir::MutMutable && !interior_unsafe {
```

Comment:
Oh, also note that WRT `&usize` this is not a change, those pointers got the `noalias` attribute before as well (the `||` got changed into a `&&`). As the comment above this change says, `noalias` is about memory dependencies rather than plain pointer equality.